provider/slof.py

```diff
@@ -0,0 +1,190 @@
+"""
+Module for providing common interface to test SLOF component.
+
+Available functions:
+ - get_boot_content: Get the specified content of SLOF by reading the serial
+                     log.
+ - wait_for_loaded: Wait for loading the SLOF.
+ - get_booted_devices: Get the device info which tried to load in the SLOF
+                       stage.
+ - verify_boot_device: Verify whether the vm is booted from the specified
+                       device.
+ - check_error: Check if there are error info in the SLOF content.
+"""
+
+import re
+import time
+import logging
+import os
+
+from virttest import utils_misc
+
+
+def get_boot_content(vm, start_pos=0, start_str='SLOF',
+                     end_str='Successfully loaded'):
+    """
+    Get the specified content of SLOF by reading the serial log.
+
+    :param vm: VM object
+    :param start_pos: start position which start to read
+    :type start_pos: int
+    :param start_str: start string
+    :type start_str: int
+    :param end_str: end string
+    :type end_str: str
+    :return: content list and next position
+    :rtype: tuple(list, int)
+    """
+    content = []
+    with open(vm.serial_console_log) as fd:
+        for pos, line in enumerate(fd):
+            if pos >= start_pos:
+                if start_str in line:
+                    content.append(line)
+                elif end_str in line:
+                    content.append(line)
+                    return content, pos + 1
+                elif content:
+                    content.append(line)
+            else:
+                continue
+        else:
+            # Reference variable 'pos' to '-1' if the serial console log is
+            # empty.
+            pos = -1
+    return None, pos + 1
+
+
+def wait_for_loaded(vm, test, start_pos=0, start_str='SLOF',
+                    end_str='Successfully loaded', timeout=300):
+    """
+    Wait for loading the SLOF.
+
+    :param vm: VM object
+    :param test: kvm test object
+    :param start_pos: start position which start to read
+    :type start_pos: int
+    :param start_str: start string
+    :type start_str: int
+    :param end_str: end string
+    :type end_str: str
+    :param timeout: time out for waiting
+    :type timeout: float
+    :return: content list and next position
+    :rtype: tuple(list, int)
+    """
+    file_timeout = 30
+    if not utils_misc.wait_for(lambda: os.path.isfile(vm.serial_console_log),
+                               file_timeout):
+        test.error('No found serial log in %s sec.' % file_timeout)
+
+    end_time = float(timeout) + time.time()
```

Comment:
I believe there is no need to convert `timeout` to float here.

Reply:
Thanks for your comment, have updated.